README.md
[![Build Status](https://travis-ci.org/yandex-money/yandex-money-sdk-ruby.svg)](https://travis-ci.org/yandex-money/yandex-money-sdk-ruby)
[![Coverage Status](https://img.shields.io/coveralls/yandex-money/yandex-money-sdk-ruby.svg)](https://coveralls.io/r/yandex-money/yandex-money-sdk-ruby)
[![Code Climate](https://codeclimate.com/github/yandex-money/yandex-money-sdk-ruby/badges/gpa.svg)](https://codeclimate.com/github/yandex-money/yandex-money-sdk-ruby)
[![Gem Version](https://badge.fury.io/rb/yandex-money-sdk.svg)](http://badge.fury.io/rb/yandex-money-sdk)
[![Dependency Status](https://gemnasium.com/yandex-money/yandex-money-sdk-ruby.svg)](https://gemnasium.com/yandex-money/yandex-money-sdk-ruby)

# Ruby Yandex.Money API SDK

## Requirements

Supported ruby versions: 1.9.3, 2.0, 2.1, jruby, rbx-2

## Links

1. Yandex.Money API page: [Ru](http://api.yandex.ru/money/),
[En](http://api.yandex.com/money/)

## Getting started

### Installation


Add this line to your Gemfile:

```ruby
gem 'yandex-money-sdk'
```

And then execute:

```
bundle
```

Or install it manually with:

```
gem install yandex-money-sdk
```

Next, require it in application:

```ruby
require 'yandex_money/api'
```


### Payments from the Yandex.Money wallet

Using Yandex.Money API requires following steps

1. Obtain token URL and redirect user's browser to Yandex.Money service.
Note: `client_id`, `redirect_uri`, `client_secret` are constants that you get,
when [register](https://sp-money.yandex.ru/myservices/new.xml) app in Yandex.Money API.

    ```ruby
    auth_url = YandexMoney::Wallet.build_obtain_token_url(
      CLIENT_ID,
      REDIRECT_URI,
      "account-info operation-history" # SCOPE
    )
    ```

2. After that, user fills Yandex.Money HTML form and user is redirected back to
`REDIRECT_URI?code=CODE`.

3. You should immediately exchange `CODE` with `ACCESS_TOKEN`.

    ```ruby
    access_token = YandexMoney::Wallet.get_access_token(
      CLIENT_ID,
      CODE,
      REDIRECT_URI
    )
    # or, if client secret defined:
    access_token = YandexMoney::Wallet.get_access_token(
      CLIENT_ID,
      CODE,
      REDIRECT_URI,
      CLIENT_SECRET
    )
    ```

    If `access_token` couldn't be obtained, `YandexMoney::ApiError` expection will be raised.


4. Now you can use Yandex.Money API.

    ```ruby
    api = YandexMoney::Wallet.new(access_token)
    account_info = api.account_info
    balance = account_info.balance # and so on

    request_options = {
        "pattern_id": "p2p",
        "to": "410011161616877",
        "amount_due": "0.02",
        "comment": "test payment comment from yandex-money-python",
        "message": "test payment message from yandex-money-python",
        "label": "testPayment",
        "test_payment": true,
        "test_result": "success"
    };
    request_result = api.request_payment(request_options)
    # check status

    process_payment = api.process_payment({
        request_id: request_result.request_id,
    })
    # check result
    if process_payment.status == "success"
      # show success page
    else
      # something went wrong
    end
    ```

### Payments from bank cards without authorization

1. Fetch instantce-id(ussually only once for every client. You can store
result in DB).

    ```ruby
    instance_id = YandexMoney::ExternalPayment.get_instance_id(CLIENT_ID)
    ```

2. Make request payment

    ```ruby
    api = YandexMoney::ExternalPayment.new(INSTANCE_ID)
    response = api.request_external_payment({
      pattern_id: "p2p",
      to: "410011285611534",
      amount_due: "1.00",
      message: "test"
    })
    if response.status == "success"
      request_id = response.request_id
    else
      # throw exception
    end
    ```

3. Process the request with process-payment. 

    ```ruby
    api = YandexMoney::ExternalPayment.new(INSTANCE_ID)
    result = api.process_external_payment({
      request_id: REQUEST_ID,
      ext_auth_success_uri: "http://example.com/success",
      ext_auth_fail_uri: "http://example.com/fail"
    })
    # process result according to docs
    ```


## Running tests

Fill values in `spec/support/constants.rb` file (example could be found in `spec/support/constants.example.rb`) and after this just run tests with `bundle exec rake` command.


CHANGELOG.md
# 1.0.4pre (December 9, 2014)

Now you could set API url through config:

```
YandexMoney.configure do |config|
  config.money_url = "https://money.yandex.ru"
  config.sp_money_url = "https://sp-money.yandex.ru"
end
```

# 1.0.3pre (November 25, 2014)

Some API changes - now exceptions raised only for 4xx and 500 error codes, changed returned values.

# 1.0.2pre (November 10, 2014)

Big refactoring of specs, they are can run without vcr cassettes with success, small fixes in README, list of operations in operations history now are OpenStruct, not hash.

# 1.0.1pre (November 7, 2014)

Change OpenStruct to RecursiveOpenStruct in all responses.

# 1.0.0pre (October 29, 2014)

Totally new API (like in other implementations), temporally removed logger, removed `spec/support/constants` from git (for travis now it encrypted).

# 0.11.0 (October 14, 2014)

Disable handling of HTTP errors inside gem in `process_payment` method.
Add support for logging.

# 0.10.0 (September 20, 2014)

Exceptions now is divided by classes, now it lives in `lib/yandex_money/exceptions.rb` for details.

# 0.9.5 (September 14, 2014)

Changed API for initialization and `obtain_token` method. Now secret key only needed in `obtain_token` method.


spec/auth_spec.rb
require "spec_helper"

describe "Application authorization flow" do
  # http://api.yandex.ru/money/doc/dg/reference/request-access-token.xml
  it "should properly initialize without client_secret" do
    VCR.use_cassette "init without client secret" do
      url = YandexMoney::Wallet.build_obtain_token_url(
        CLIENT_ID,
        REDIRECT_URI,
        "account-info operation-history"
      )
      expect(url).to start_with("https://money.yandex.ru/select-wallet.xml?requestid=")
    end
  end

  # http://api.yandex.ru/money/doc/dg/reference/request-access-token.xml
  it "should get token from code" do
    VCR.use_cassette "get token from authorization code" do
      expect(
        YandexMoney::Wallet.get_access_token(
          CLIENT_ID,
          "SOME CODE",
          REDIRECT_URI
        )
      ).to be nil
    end
  end

  it "could be initialized with token" do
    VCR.use_cassette "initialize with token" do
      api = YandexMoney::Wallet.new(ACCESS_TOKEN)
      expect(api.account_info.account).to eq WALLET_NUMBER
    end
  end
end


spec/cards_spec.rb
require "spec_helper"

describe "Payments from bank cards without authorization" do
  before :all do
    VCR.use_cassette "obtain token for payments from bank cards without authorization" do
    end
  end
  it "should fail when try to register an instance of application without connected market" do
    VCR.use_cassette "get instance id fail" do
      expect(
        YandexMoney::ExternalPayment.get_instance_id(nil).status
      ).to eq "refused"
    end
  end

  it "should register an instance of application" do
    VCR.use_cassette "get instance id success" do
      expect(
        YandexMoney::ExternalPayment.get_instance_id(CLIENT_ID)
                                    .instance_id.length
      ).to eq 64
    end
  end

  it "should request external payment" do
    VCR.use_cassette "request external payment" do
      instance_id = YandexMoney::ExternalPayment.get_instance_id(CLIENT_ID)
      @api = YandexMoney::ExternalPayment.new(instance_id)
      expect(@api.request_external_payment({
        pattern_id: "p2p",
        to: "410011285611534",
        amount_due: "1.00",
        message: "test"
      }).status).to eq("success")
    end
  end

  it "should process external payment" do
    VCR.use_cassette "process external payment" do
      instance_id = YandexMoney::ExternalPayment.get_instance_id(CLIENT_ID)
      @api = YandexMoney::ExternalPayment.new(instance_id)
      request_id = @api.request_external_payment(
        pattern_id: "p2p",
        to: "410011285611534",
        amount_due: "1.00",
        message: "test"
      ).request_id
      expect(@api.process_external_payment({
        request_id: request_id,
        ext_auth_success_uri: "http://127.0.0.1:4567/success",
        ext_auth_fail_uri: "http://127.0.0.1:4567/fail"
      }).status).to eq("ext_auth_required")
    end
  end
end


spec/config_spec.rb
require "spec_helper"

describe "YandexMoney config" do
  after :all do
    YandexMoney.reset_config
  end

  let(:default_money_url) { "https://money.yandex.ru" }
  let(:default_sp_money_url) { "https://sp-money.yandex.ru" }

  it "should have methods: config, configure, reset_config" do
    yandex = class_double("YandexMoney")
    expect(yandex).to receive_messages([:config, :configure, :reset_config])
    yandex.config & yandex.configure & yandex.reset_config
  end

  it "should yield block in configure" do
    expect {|b| YandexMoney.configure(&b)}.to yield_control
  end

  it "should return a YandexMoney::Config instance" do
    expect(YandexMoney.config).to be_a_kind_of(YandexMoney::Config)
  end

  it "should reset_config" do
    YandexMoney.configure { |c| c.money_url = "http://example1.com" }
    YandexMoney.reset_config
    expect(YandexMoney.config.money_url).to eq(default_money_url)
  end

  it "should have correct defaults" do
    expect(YandexMoney.config.money_url).to eq(default_money_url)
    expect(YandexMoney.config.sp_money_url).to eq(default_sp_money_url)
  end

  it "can write config url directly" do
    YandexMoney.config.money_url = "http://example3.com"
    expect(YandexMoney.config.money_url).to eq "http://example3.com"
  end

  it "should correct configure hosts addresses" do
    YandexMoney.configure do |config|
      config.money_url = "http://example1.com"
      config.sp_money_url = "http://example2.com"
    end
    expect(YandexMoney.config.money_url).to eq "http://example1.com"
    expect(YandexMoney.config.sp_money_url).to eq "http://example2.com"
  end

end


lib/yandex_money/exceptions.rb
module YandexMoney
  class Error < StandardError; end

  class ErrorWithMessage < Error
    def initialize(error)
      @err = error
    end

    def message
      if @err["www-authenticate"]
        "#{@err.msg} - #{@err["www-authenticate"]}"
      else
        @err.msg
      end
    end
  end

  class InsufficientScopeError < ErrorWithMessage; end

  class InvalidRequestError < ErrorWithMessage; end

  class UnauthorizedError < ErrorWithMessage; end

  class ServerError < Error; end

  class FieldNotSetError < Error
    def initialize(field)
      @field = field
    end

    def message
      "Field '#{@field}' not set!"
    end
  end

  class ApiError < Error
    def initialize(msg)
      @msg = msg
    end

    def message
      @msg.gsub(/_/, " ").capitalize
    end
  end

  class AcceptTransferError < ApiError
    def initialize(msg, attemps = nil)
      @msg = msg
      @attemps = attemps
    end

    def message
      if @attemps
        super + ", attemps available: #{@attemps}"
      else
        super
      end
    end
  end

  class ExternalPaymentProgressError < ApiError
    def initialize(msg, next_retry)
      @msg = msg
      @next_retry = next_retry
    end

    def message
      super + ", next_retry: #{@next_retry}"
    end
  end
end


lib/yandex_money/external_payment.rb
module YandexMoney
  # Payments without auth
  class ExternalPayment
    def initialize(instance_id)
      @instance_id = instance_id
    end

    def self.get_instance_id(client_id)
      request = send_external_payment_request("/api/instance-id", client_id: client_id)
      RecursiveOpenStruct.new request.parsed_response
    end

    def request_external_payment(payment_options)
      payment_options[:instance_id] = @instance_id
      request = self.class.send_external_payment_request("/api/request-external-payment", payment_options)
      RecursiveOpenStruct.new request.parsed_response
    end

    def process_external_payment(payment_options)
      payment_options[:instance_id] = @instance_id
      request = self.class.send_external_payment_request("/api/process-external-payment", payment_options)
      RecursiveOpenStruct.new request.parsed_response
    end

    private

    def self.send_external_payment_request(uri, options)
      request = HTTParty.post(uri, base_uri: YandexMoney.config.money_url, headers: {
        "Content-Type" => "application/x-www-form-urlencoded"
      }, body: options)
      case request.response.code
      when "400" then raise YandexMoney::InvalidRequestError.new request.response
      when "401" then raise YandexMoney::UnauthorizedError.new request.response
      when "403" then raise YandexMoney::InsufficientScopeError request.response
      when "500" then raise YandexMoney::ServerError
      else
        request
      end
    end

  end
end


lib/yandex_money/wallet.rb
require "httparty"
require "recursive-open-struct"

module YandexMoney
  # Payments from YandexMoney wallet
  class Wallet
    include HTTParty
    base_uri YandexMoney.config.money_url
    default_timeout 30

    attr_accessor :token

    def initialize(token)
      @token = token
    end

    # obtains account info
    def account_info
      RecursiveOpenStruct.new send_request("/api/account-info", recurse_over_arrays: true).parsed_response
    end

    # obtains operation history
    def operation_history(options=nil)
      history = RecursiveOpenStruct.new(
        send_request("/api/operation-history", options).parsed_response
      )
      history.operations = history.operations.map do |operation|
        RecursiveOpenStruct.new operation
      end
      history
    end

    # obtains operation details
    def operation_details(operation_id)
      request = send_request("/api/operation-details", operation_id: operation_id)
      RecursiveOpenStruct.new request.parsed_response
    end

    # basic request payment method
    def request_payment(options)
      send_payment_request("/api/request-payment", options)
    end

    # basic process payment method
    def process_payment(options)
      send_payment_request("/api/process-payment", options)
    end

    def incoming_transfer_accept(operation_id, protection_code = nil)
      uri = "/api/incoming-transfer-accept"
      if protection_code
        request_body = {
          operation_id: operation_id,
          protection_code: protection_code
        }
      else
        request_body = { operation_id: operation_id }
      end
      RecursiveOpenStruct.new send_request("/api/incoming-transfer-accept", request_body)
    end

    def incoming_transfer_reject(operation_id)
      RecursiveOpenStruct.new send_request("/api/incoming-transfer-reject", operation_id: operation_id)
    end

    def self.build_obtain_token_url(client_id, redirect_uri, scope)
      uri = "#{YandexMoney.config.sp_money_url}/oauth/authorize"
      options = {
        client_id: client_id,
        response_type: "code",
        redirect_uri: redirect_uri,
        scope: scope
      }
      HTTParty.post(uri, body: options).request.path.to_s
    end

    def self.get_access_token(client_id, code, redirect_uri, client_secret=nil)
      uri = "#{YandexMoney.config.sp_money_url}/oauth/token"
      options = {
        code: code,
        client_id: client_id,
        grant_type: "authorization_code",
        redirect_uri: redirect_uri
      }
      options[:client_secret] = client_secret if client_secret
      response = HTTParty.post(uri, body: options).parsed_response
      response["access_token"]
    end

    protected

    def send_request(uri, options = nil)
      request = self.class.post(uri, headers: {
        "Authorization" => "Bearer #{@token}",
        "Content-Type" => "application/x-www-form-urlencoded"
      }, body: options)
      case request.response.code
      when "400" then raise YandexMoney::InvalidRequestError.new request.response
      when "401" then raise YandexMoney::UnauthorizedError.new request.response
      when "403" then raise YandexMoney::InsufficientScopeError.new request.response
      when "500" then raise YandexMoney::ServerError
      else
        request
      end
    end

    def send_payment_request(uri, options)
      request = send_request(uri, options)
      response = RecursiveOpenStruct.new request.parsed_response
      response
    end
  end
end
